### PR TITLE
Implemented ActiveRecord parameter binding #9890

### PR DESCRIPTION
**Pull request by SamMousa**
Tests need some refactoring.

### CODE REVIEW

File: framework/web/Controller.php [hunk 1]
```diff
@@ -81,6 +81,56 @@ public function bindActionParams($action, $params)
             if ($class !== null && (!array_key_exists($name, $params) || !$params[$name] instanceof $className)) {
                 if (Yii::$app->has($name) && ($obj = Yii::$app->get($name)) instanceof $className) {
                     $args[] = $actionParams[$name] = $obj;
+                } elseif (isset(class_implements($className)['yii\db\ActiveRecordInterface'])) {
```
Comment (samdark) [inline framework/web/Controller.php]: No need for `isset`.

File: framework/web/Controller.php [hunk 2]
```diff
@@ -81,6 +81,56 @@ public function bindActionParams($action, $params)
...
+                     * and if the type hint hints to an ActiveRecord class.
+                     * If so, it checks if it the value is a valid / safe for passing into the query.
+                     * (Safe meaning it will only search on columns using the primary key(s))
+                     */
+
+                    if (isset($params[$name])) {
+                        if (count($className::primaryKey()) == 1 && is_scalar($params[$name])
+                            || (
+                                is_array($params[$name])
+                                && count($className::primaryKey()) == count($params[$name])
```
Comment (samdark) [inline framework/web/Controller.php]: `===` could be used

File: framework/web/Controller.php [hunk 3]
```diff
@@ -81,6 +81,56 @@ public function bindActionParams($action, $params)
...
+                    /**
+                     * This is a pretty complex guard.
+                     * It checks if a request parameter with the same name as the action parameter is set,
+                     * and if the type hint hints to an ActiveRecord class.
+                     * If so, it checks if it the value is a valid / safe for passing into the query.
+                     * (Safe meaning it will only search on columns using the primary key(s))
+                     */
+
+                    if (isset($params[$name])) {
+                        if (count($className::primaryKey()) == 1 && is_scalar($params[$name])
```
Comment (samdark) [inline framework/web/Controller.php]: `===` could be used

File: framework/web/Controller.php (hunk 1 above)
Comment (SamMousa) [inline framework/web/Controller.php]: Won't we get undefined index errors if we omit it?

Comment (samdark) [inline framework/web/Controller.php]: Oh. What a weird construct. It is correct. Ignore my comment.

**Comment by samdark** [comment]
#9890

**Comment by mdmunir** [comment]
IMO its too complex for DI and also not compatible with current pattern.
If we accept parameter hinting for AR, what parameter name should be? Use `actionView(Project $project)` or `actionView(Project $id)`?
If we use `$project` as parameter name, its mean its not compatible with `yii\rest\UrlRule`. If we use `$id` as parameter name, its not good looking for object variable name.

**Comment by samdark** [comment]
I agree that it's a bit too much for DI itself but:

1. The feature is implemented on controller level so it's controller-specific and that makes it OK in this regard.
2. The same thing exists in RoR. While it's very "magical" people tend to love it.

The naming here is very tough thing. I don't like both `$id` and `$project` same as @mdmunir.

**Comment by SamMousa** [comment]
I see what you mean; of course don't have to necessarily use the same names.

The way we do work is quite simple, we first check if a type hinted parameter which should be named $project, has an AR class.
If that is the case then we can try to match request parameters to that, by using the primary key name as parameter name.

This has an added advantage of creating less magic because the original parameter is still available.

For example:

        public function actionRead(Project $project, $id)

We could populate both parameters if we wanted to, regardless if we do this, we can throw proper errors even in case of:

        public function actionRead(Project $project)

We could throw 400, $id is a required parameter (if that is the name of the primary key for Project).

Waht do you guys think, @samdark, @mdmunir ?

**Comment by samdark** [comment]
As far as I see, technically parameter could be named whatever you choose so in docs it's probably a good idea to use `$project` instead of `$id`. That's a bit more good looking in controllers.

**Comment by samdark** [comment]
I don't like an idea to pass `$id` separately. It doesn't make sense like that.

**Comment by SamMousa** [comment]
I think you misunderstood me. I mean to use the pk name as parameter name
in the URL regardless of the name in the code.
And that would allow to pass the actual id value because it uses the
current parameter binding code.

We could even make the key variable name configurable in the controller so
you can override it easily.

**Comment by samdark** [comment]
What if I need two models at the same time for an action?

**Comment by SamMousa** [comment]
Several options: use other notation like projectId, only type hint one and
don't support this magic for 2 models.

Let me read up on how RoR does it.

**Comment by samdark** [comment]
Hmm... probably it wasn't rails.

**Comment by SamMousa** [comment]
Okay, so currently multiple model loading would not be natively supported.
The question is, does that really matter?
* In most cases there is 1 "main" model, so in these cases this feature helps.
* You still have access to AR via the normal way, so nothing prevents you from using it without DI / automatic querying.
* We could add more intelligent parsing later, for example:
-- Case multiple params have an AR typehint, and they have the same primaryKey name, we:
1. Check if the query string contains parameters like ModelPrimaryKey first, that way we can map those to the appropriate models.
2. Check the remaining AR model parameters to see if we still have duplicate primary keys remaining, if so, throw an exception.

Anyway, I believe that it is not necessary at this point to support 100% of the cases, but instead just supporting the single model use case and then waiting for community feedback would be great.

That way we have controlled change.

**Comment by pana1990** [comment]
this feature is include in Laravel 5.2 https://laravel.com/docs/5.0/routing#route-model-binding

I suggest include getRouteKeyName in ActiveRecord as laravel

**Comment by klimov-paul** [comment]
Unfortunally we can not accept this feature in the core.
Still perhaps you will be able to realse it a standalone extension.

You still should be able to extend `yii\web\Controller::bindActionParams()` and adjust it logic:

```php
public function bindActionParams($action, $params)
{
      $args = parent::bindActionParams($action, $params);
     
      // place extra AR processing logic here

      return $args;
}
```

Thus you should be able to create a custom controller class or trait.

**Pull request closed without merging**